File: analyzer/nodejs/src/sources/integrity.ts
/** Inline module integrity — the `#sha256-<base64url>` fragment carried on a
 *  remote import ref. Browser-safe: uses Web Crypto (`crypto.subtle`) and
 *  `btoa`, both globals in Node and the browser. No Node built-ins.
 *
 *  The fragment is authoritative across every transport: a source's `read()`
 *  hashes the fetched bytes and compares against it before the manifest is
 *  parsed or cached. A mismatch is a terminal error — never a cache miss. */

/** Only a `#<alg>-<base64url>` suffix is treated as integrity; other `#`
 *  fragments (rare in module refs) pass through untouched. `sha256` is the
 *  only algorithm accepted today; the prefix leaves room to migrate. */
const INTEGRITY_FRAGMENT = /#(sha256-[A-Za-z0-9_+/=-]+)$/;

/** A failed integrity/tamper check — always terminal, never best-effort. A
 *  distinct type so a caller doing best-effort network handling (e.g. the
 *  bundle extractor warning-and-skipping on a fetch blip) can still let a
 *  tamper error propagate rather than swallow it. */
export class IntegrityError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IntegrityError";
  }
}

/** Split a trailing integrity fragment off a ref/URL. Returns the bare ref in
 *  `base` (safe to build fetch URLs and cache paths from) and the fragment in
 *  `integrity` (e.g. `sha256-<base64url>`), or `undefined` when absent. */
export function splitIntegrity(ref: string): { base: string; integrity?: string } {
  const match = ref.match(INTEGRITY_FRAGMENT);
  if (!match) return { base: ref };
  return { base: ref.slice(0, match.index), integrity: match[1] };
}

/** Attach an integrity hash to a ref as a `#<alg>-...` fragment. No-op when the
 *  hash is absent/non-string or the ref already carries a fragment (an
 *  author-authored pin is never overwritten). Inverse of `splitIntegrity`;
 *  used to fold the object form's `integrity:` sibling into the source string. */
export function foldIntegrity(source: string, integrity: unknown): string {
  return typeof integrity === "string" && !source.includes("#")
    ? `${source}#${integrity}`
    : source;
}

function toBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/** Normalize an encoded digest to unpadded base64url so a standard-base64 or
 *  padded input still compares equal to our canonical form. */
function normalizeDigest(value: string): string {
  return value.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/** SHA-256 of `bytes` as unpadded base64url — the canonical inline-hash form. */
export async function sha256Base64Url(bytes: Uint8Array): Promise<string> {
  // Copy into a plain ArrayBuffer: a Uint8Array may be backed by a
  // SharedArrayBuffer, which `crypto.subtle.digest` does not accept.
  const buffer = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(buffer).set(bytes);
  const digest = await crypto.subtle.digest("SHA-256", buffer);
  return toBase64Url(new Uint8Array(digest));
}

/** Hash `bytes` and compare against `integrity` (`<alg>-<digest>`). Throws a
 *  terminal error on mismatch or an unsupported algorithm. `describe` names the
 *  artifact in the error (e.g. the module ref) so the failure is actionable. */
export async function verifyIntegrity(
  bytes: Uint8Array,
  integrity: string,
  describe: string,
): Promise<void> {
  const dash = integrity.indexOf("-");
  const algorithm = dash > 0 ? integrity.slice(0, dash) : "";
  const expected = dash > 0 ? integrity.slice(dash + 1) : "";
  if (algorithm !== "sha256") {
    throw new IntegrityError(
      `Unsupported integrity algorithm '${algorithm || integrity}' for ${describe}. ` +
        `Only sha256 is supported (sha256-<base64url>).`,
    );
  }
  const actual = await sha256Base64Url(bytes);
  if (actual !== normalizeDigest(expected)) {
    throw new IntegrityError(
      `Integrity check failed for ${describe}: expected sha256-${normalizeDigest(expected)}, ` +
        `got sha256-${actual}. The fetched bytes do not match the recorded hash — ` +
        `the module may have been tampered with or republished.`,
    );
  }
}

/** The single verified network read for remote manifests: fetch `fetchUrl`,
 *  verify the raw bytes against `integrity` (when pinned), and return both the
 *  bytes and the decoded text. The one choke point every network `ManifestSource`
 *  routes through, so verification cannot drift between them. `describe` names
 *  the artifact in error messages. */
export async function verifiedFetch(
  fetchUrl: string,
  integrity: string | undefined,
  describe: string,
): Promise<{ bytes: Uint8Array; text: string }> {
  const response = await fetch(fetchUrl);
  if (!response.ok) {
    throw new Error(
      `Failed to fetch manifest ${describe}: ${response.status} ${response.statusText} (${fetchUrl})`,
    );
  }
  const bytes = new Uint8Array(await response.arrayBuffer());
  if (integrity) await verifyIntegrity(bytes, integrity, describe);
  return { bytes, text: new TextDecoder().decode(bytes) };
}

File: analyzer/rust/src/sources/integrity.rs
//! Inline module integrity — the `#sha256-<base64url>` fragment carried on a
//! remote import ref. Mirrors `../../../nodejs/src/sources/integrity.ts`.
//!
//! The fragment is authoritative across every transport: a source hashes the
//! fetched bytes and compares against it before the manifest is parsed or
//! cached. A mismatch is a terminal error — never a cache miss.
//!
//! Only what the Rust kernel consumes is here. `verifiedFetch` and its
//! object-storage sniff belong to the Node analyzer's HTTP source, which has no
//! Rust counterpart; `foldIntegrity` and `isCanonicalIntegrity` serve the import
//! object form and pin writing, neither of which this kernel implements.

use sha2::{Digest, Sha256};

const SHA256_PREFIX: &str = "sha256-";

/// A failed integrity/tamper check — always terminal, never best-effort.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{message}")]
pub struct IntegrityError {
    pub message: String,
}

impl IntegrityError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

/// A ref with its trailing integrity fragment split off.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SplitRef<'a> {
    /// The bare ref, safe to build fetch URLs and cache paths from.
    pub base: &'a str,
    /// The fragment, e.g. `sha256-<base64url>`, when present.
    pub integrity: Option<&'a str>,
}

/// Split a trailing integrity fragment off a ref/URL. Only a
/// `#sha256-<base64url>` suffix is integrity; any other `#` fragment passes
/// through untouched. Tolerates the padded / standard-base64 spellings
/// [`verify_integrity`] normalizes.
pub fn split_integrity(reference: &str) -> SplitRef<'_> {
    // The fragment's alphabet excludes `#`, so only the last `#` can start one.
    if let Some(hash) = reference.rfind('#') {
        let fragment = &reference[hash + 1..];
        let digest = fragment.strip_prefix(SHA256_PREFIX).unwrap_or("");
        let is_integrity = !digest.is_empty()
            && digest
                .chars()
                .all(|c| c.is_ascii_alphanumeric() || matches!(c, '_' | '+' | '/' | '=' | '-'));
        if is_integrity {
            return SplitRef {
                base: &reference[..hash],
                integrity: Some(fragment),
            };
        }
    }
    SplitRef {
        base: reference,
        integrity: None,
    }
}

const BASE64URL: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

fn to_base64url(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len().div_ceil(3) * 4);
    for chunk in bytes.chunks(3) {
        let b = [
            chunk[0],
            chunk.get(1).copied().unwrap_or(0),
            chunk.get(2).copied().unwrap_or(0),
        ];
        let n = (u32::from(b[0]) << 16) | (u32::from(b[1]) << 8) | u32::from(b[2]);
        let symbols = chunk.len() + 1;
        for i in 0..symbols {
            out.push(BASE64URL[((n >> (18 - 6 * i)) & 0x3f) as usize] as char);
        }
    }
    out
}

/// Normalize an encoded digest to unpadded base64url so a standard-base64 or
/// padded input still compares equal to the canonical form.
fn normalize_digest(value: &str) -> String {
    value
        .replace('+', "-")
        .replace('/', "_")
        .trim_end_matches('=')
        .to_string()
}

/// SHA-256 of `bytes` as unpadded base64url — the canonical inline-hash form.
pub fn sha256_base64url(bytes: &[u8]) -> String {
    to_base64url(&Sha256::digest(bytes))
}

/// Hash `bytes` and compare against `integrity` (`<alg>-<digest>`). `describe`
/// names the artifact in the error so the failure is actionable.
pub fn verify_integrity(bytes: &[u8], integrity: &str, describe: &str) -> Result<(), IntegrityError> {
    let (algorithm, expected) = match integrity.find('-') {
        Some(dash) if dash > 0 => (&integrity[..dash], &integrity[dash + 1..]),
        _ => ("", ""),
    };
    if algorithm != "sha256" {
        let named = if algorithm.is_empty() { integrity } else { algorithm };
        return Err(IntegrityError::new(format!(
            "Unsupported integrity algorithm '{named}' for {describe}. Only sha256 is supported (sha256-<base64url>)."
        )));
    }
    let actual = sha256_base64url(bytes);
    let expected = normalize_digest(expected);
    if actual != expected {
        return Err(IntegrityError::new(format!(
            "Integrity check failed for {describe}: expected sha256-{expected}, got sha256-{actual}. \
             The fetched bytes do not match the recorded hash — the module may have been tampered with or republished."
        )));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Mirrors `analyzer/nodejs/tests/integrity.test.ts` § splitIntegrity.
    #[test]
    fn splits_only_an_integrity_fragment() {
        assert_eq!(
            split_integrity("oci://ghcr.io/telorun/console@0.9.0#sha256-AAAA"),
            SplitRef {
                base: "oci://ghcr.io/telorun/console@0.9.0",
                integrity: Some("sha256-AAAA"),
            }
        );
        assert_eq!(
            split_integrity("oci://ghcr.io/telorun/console@0.9.0").integrity,
            None
        );
        assert_eq!(
            split_integrity("http://x/a.yaml#section"),
            SplitRef {
                base: "http://x/a.yaml#section",
                integrity: None,
            }
        );
    }

    /// Mirrors § verifyIntegrity: a match passes, a padded standard-base64
    /// spelling of the same digest passes, a mismatch and an unknown algorithm
    /// are terminal and name the artifact.
    #[test]
    fn verifies_a_pin() {
        let bytes = b"abc";
        // Node's `createHash("sha256").update("abc").digest("base64url")`.
        let pin = format!("sha256-{}", sha256_base64url(bytes));
        assert_eq!(pin, "sha256-ungWv48Bz-pBQUDeXa4iI7ADYaOWF3qctBD_YfIAFa0");
        verify_integrity(bytes, &pin, "ref").unwrap();
        verify_integrity(bytes, "sha256-ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0=", "ref").unwrap();

        let mismatch = verify_integrity(b"tampered", "sha256-AAAA", "oci://ghcr.io/aws/s3@1.0.0")
            .unwrap_err();
        assert!(
            mismatch
                .message
                .starts_with("Integrity check failed for oci://ghcr.io/aws/s3@1.0.0"),
            "{mismatch}"
        );
        let algorithm = verify_integrity(b"x", "md5-AAAA", "ref").unwrap_err();
        assert!(algorithm.message.starts_with("Unsupported integrity algorithm 'md5'"), "{algorithm}");
    }
}
